Update Rollbar

Tkael opened the pull request:
Reduces web traffic when idle, adds telemetry data from prior eddi.log entries.

Tkael committed:
Update Rollbar to latest version, update `configureRollbar` method. Add a method for enabling or disabling Rollbar via the Utilities class.

Tkael committed:
Remove Rollbar from the Test class.

Tkael committed:
Make Rollbar's unique ID unique to the app run rather than unique to the installation.

Tkael committed:
Tidy up

Tkael committed:
Add telemetry reporting.

Tkael committed:
Remove unused Restsharp API checks for uniqueness.

Tkael committed:
Local log should report exactly what was sent to Rollbar (rather than the unfiltered and unredacted version).

Tkael committed:
Telemetry tweaks.

richardbuckle commented on Utilities/Logging.cs:
```diff
@@ -69,6 +70,7 @@ private static void log(ErrorLevel errorlevel, string data, string method, strin
                     // Failed; can't do anything about it as we're in the logging code anyway
                 }
             }
+            WriteTelemetry($"{shortPath}:{method}", data);
```
Isn't this sending for *all* logging levels, not just the ones we want to send?

Tkael replied:
Yes, this will provide context from the info and debug level logs written immediately prior to the error. I think this context will be useful.
Data is subject to the same redactions and filters as any other data sent to Rollbar and is not sent until an error is sent.

GioVAX replied:
If an exception is thrown before `shortPath` is initialized, we might get to line 73 with a null value, and I do not remember how `$` behaves with null values.

Tkael replied:
It behaves as an empty string. Confirmed with the following test:
```cs
        [TestMethod]
        public void TestTemp()
        {
            string shortPath = null;
            Assert.AreEqual("", $"{shortPath}");
        }
```

GioVAX replied:
SG

richardbuckle commented on EDDI/App.xaml.cs:
```diff
@@ -103,12 +103,9 @@ public static void StartRollbar()
-            }
-            _Rollbar.configureRollbar(Eddi.Properties.Settings.Default.uniqueID);
+            // Generate an id unique to this app run for bug tracking
+            if (!string.IsNullOrEmpty(Eddi.Properties.Settings.Default.uniqueID)) { Eddi.Properties.Settings.Default.uniqueID = null; }
```
1. Is this desirable?
2. If so, does this mean we can do away with `Eddi.Properties.Settings.Default.uniqueID`?

Tkael replied:
1. I believe so.
2. Yes.

richardbuckle commented on Utilities/Logging.cs:
```diff
@@ -112,18 +114,38 @@ public static void incrementLogs()
-            Dictionary<string, object> thisData = PrepRollbarData(ref data);
-            if (thisData != null)
+            Dictionary<string, object> preppedData = PrepRollbarData(ref originalData);
+            if (originalData is null || preppedData != null)
```
Sorry, I don't follow why we've added `originalData is null` here.

Tkael replied:
As originally written, a message without a data package would resolve to false here and wouldn't be sent to Rollbar. This removes the requirement to include a data package with the message.

richardbuckle commented on Utilities/Logging.cs:
```diff
@@ -282,13 +301,16 @@ public static void configureRollbar(string uniqueId)
                     Root = "/"
                 },
                 MaxReportsPerMinute = 1,
+                IpAddressCollectionPolicy = IpAddressCollectionPolicy.DoNotCollect,
```
This I would keep regardless of what we do with the rest.

richardbuckle commented on Utilities/Logging.cs:
```diff
@@ -298,124 +320,13 @@ public static void ExceptionHandler(Exception exception)
-                Logging.Info("Reporting unhandled exception, anonymous ID " + RollbarLocator.RollbarInstance.Config.Person.Id + ":" + exception);
-                RollbarLocator.RollbarInstance.Error(exception, trace);
-            }
+            Logging.Info("Reporting unhandled exception, anonymous ID " + RollbarLocator.RollbarInstance.Config.Person.Id + ":" + exception);
```
Did we intend to lose the uniqueness test here?

Tkael replied:
We were only testing for uniqueness on info level messages and we were only sending error level messages. Consequently, we were already not testing any messages for uniqueness.

richardbuckle commented on Utilities/Utilities.csproj:
```diff
@@ -54,25 +54,139 @@
     <Reference Include="MathNet.Numerics, Version=4.7.0.0, Culture=neutral, processorArchitecture=MSIL">
       <HintPath>..\packages\MathNet.Numerics.4.7.0\lib\net461\MathNet.Numerics.dll</HintPath>
     </Reference>
+    <Reference Include="Microsoft.Data.Sqlite, Version=2.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60, processorArchitecture=MSIL">
```
Wow a huge load of MS stuff added to Utilities that I did not expect. Are we sure about all this?

Tkael replied:
It was more than I expected too. It it weren't all MS stuff, I'd be much more reluctant.

richardbuckle commented on Utilities/packages.config:
```diff
@@ -2,9 +2,53 @@
 <packages>
   <package id="GitInfo" version="2.0.10" targetFramework="net471" developmentDependency="true" />
   <package id="MathNet.Numerics" version="4.7.0" targetFramework="net471" />
+  <package id="Microsoft.CSharp" version="4.4.0" targetFramework="net48" />
```
Again a huge load of MS stuff added to Utilities that I did not expect. Are we sure about all this?

Tkael replied:
It was more than I expected too. It it weren't all MS stuff, I'd be much more reluctant. 

If we want to consider an alternative telemetry service, I still have an open source license ready to go with Bugsnag.

richardbuckle reviewed (changes requested):
In general I'm worried that we are fixing something that isn't broken here, just because one random blowhard had a hissy fit. 

There is definitely stuff worth keeping here, but we must not get into the habit of going off and doing work without first agreeing it with the team (ref `docs/Contributing.md`).

As such, I have to say that I have considerable misgivings here.

richardbuckle commented:
Adding @GioVAX and @Hoodathunk for their opinions.

I won't lie to you: I'm nervous that we are fixing something that isn't broken for no good reason.

Tkael commented:
We're significantly reducing the packet size for server-client handshakes (was up to 1000 bytes on some packets, is consistently 50-60 bytes), we're improving the quality of the reports that we send/receive, we're further anonymizing the data that we're sending, and we're removing dead/unused code.

GioVAX commented on Tests/TestBase.cs:
```diff
@@ -11,7 +10,7 @@ public class TestBase
         {
             // Prevent telemetry data from being reported based on test results
-            RollbarLocator.RollbarInstance.Config.Enabled = false;
+            Utilities._Rollbar.EnableDisable(false);
```
Can we rename this method to something more revealing?
E.g. SendTelemetryData

Tkael replied:
Agreed.

Tkael committed:
Use a more revealing name for our Telemetry toggle.

Tkael committed:
Remove unused variables.

richardbuckle commented:
I am OK with the code changes but I want to understand what caused the massive cascade of MS dependencies. If it can be explained then I'd be happier.

Tkael commented:
Rollbar's seen a lot of development since we first added it. We're moving from [version 1.1.1 to version 3.12.0]((https://github.com/rollbar/Rollbar.NET/compare/v1.1.1...v3.12.0) ). The full release history is on Github [here](https://github.com/rollbar/Rollbar.NET/tags). 

Here's what I've been able to find out about some of the new dependencies which are listed for .NET framework version 4.8:
- [3.0.3](https://github.com/rollbar/Rollbar.NET/releases/tag/v3.0.3) Added `Microsoft.Extensions.Configuration`, `Microsoft.Extensions.Configuration.Binder`, and `Microsoft.Extensions.Configuration.Json` dependencies. Seems to have been added to resolve [this issue](https://github.com/rollbar/Rollbar.NET/issues/293).
- [3.6.2](https://github.com/rollbar/Rollbar.NET/releases/tag/v3.6.2) Added `Microsoft.EntityFrameworkCore.Sqlite` dependency. This change allows Rollbar to [store errors](https://github.com/rollbar/Rollbar.NET/issues/424) in an sql database until it is able to report them to the server.
